argon2/src/memory_view.rs

```diff
@@ -0,0 +1,118 @@
+//! This module encapsulates the unsafe code for accessing memory in parallel.
+
+// TODO(tarcieri): safe parallel implementation
+// See: https://github.com/RustCrypto/password-hashes/issues/154
+
+use crate::{instance::SYNC_POINTS, Block};
+
+use core::slice;
+
+#[cfg(feature = "parallel")]
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+#[derive(Debug)]
+pub struct MemoryView<'a> {
+    memory: &'a mut [Block],
+    slice: usize,
+    lane_count: usize,
+    segment_length: usize,
+
+    #[cfg(feature = "parallel")]
+    cur_lane: AtomicUsize,
+
+    #[cfg(not(feature = "parallel"))]
+    cur_lane: usize,
+}
+
+impl<'a> MemoryView<'a> {
+    pub fn new(memory: &'a mut [Block], slice: usize, lane_count: usize) -> Self {
+        assert!(slice < SYNC_POINTS);
+        assert_ne!(lane_count, 0);
+
+        let segment_count = lane_count * SYNC_POINTS;
+        assert!(memory.len() >= segment_count);
+        assert_eq!(memory.len() % segment_count, 0);
+
+        let segment_length = memory.len() / segment_count;
+
+        Self {
+            memory,
+            slice,
+            lane_count,
+            segment_length,
+
+            #[cfg(feature = "parallel")]
+            cur_lane: AtomicUsize::new(0),
+
+            #[cfg(not(feature = "parallel"))]
+            cur_lane: 0,
+        }
+    }
+
+    pub fn lane(&self, lane: usize) -> (&'a [Block], &'a [Block]) {
+        assert!(lane < self.lane_count);
+
+        let base_ptr = self.memory.as_ptr();
+
+        unsafe {
+            let before_ptr = base_ptr.add(lane * self.lane_length());
+            let before_len = self.slice * self.segment_length;
+            let before_segment = slice::from_raw_parts(before_ptr, before_len);
+
+            let offset = before_len + self.segment_length;
+            let after_ptr = before_ptr.add(offset);
+            let after_len = self.lane_length() - offset;
+            let after_segment = slice::from_raw_parts(after_ptr, after_len);
+
+            (before_segment, after_segment)
+        }
+    }
+
+    #[cfg(feature = "parallel")]
+    pub fn next_segment(&self) -> Option<(&'a mut [Block], usize)> {
+        self.cur_lane
+            .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |lane| {
+                if lane < self.lane_count {
+                    Some(lane + 1)
+                } else {
+                    None
+                }
+            })
+            .map(|lane| {
+                let base_ptr = self.memory.as_ptr();
+
+                let segment = unsafe {
+                    let start_ptr =
+                        base_ptr.add(lane * self.lane_length() + self.slice * self.segment_length);
+                    slice::from_raw_parts_mut(start_ptr as *mut Block, self.segment_length)
+                };
```

Comment:
This looks as if it's mutably aliasing a region of memory declared immutable. If so, that's unsound, given this method has a safe API.

Seems like a showstopper unless it can be addressed.

Reply:
`base_ptr` is indeed a const pointer because we cannot call `as_mut_ptr` since the mutable reference to the memory is behind an immutable reference to `self`.
However, it is sound to cast `start_ptr` to a mut pointer because:
- `MemoryView` holds a mutable reference to the memory and therefore has exclusive access.
- The `cur_lane` field is incremented every time a mutable slice is returned and is not used elsewhere, so the slices `next_segment` returns never overlap.
- The `slice` field is never modified, and the `lane` method specifically avoids the parts of the memory that `next_segment` returns.